### PR TITLE
Fix multiple optionals

### PR DESCRIPTION
**Pull request by RickMoynihan**
# TPX: matcha: fix multiple optionals

Issue: https://github.com/Swirrl/matcha/issues/21

In short: handling of `optional` clauses in matcha queries was returning incorrect results (depending on order). Root cause of that was the use of `core.logic/conda` where `core.logic/conde` should be used. But that change necessitates extra processing of the the raw core logic results (for both `select` and `build` macros). 

### `build`

We infer which projected vars in the constructed map are optional. Knowing that, and having a solution we either set the (top level!) keys to the value from the solution (a set, if multiple values were returned by `core.logic`) or remove those keys.

We also rewrite the query by extracting the `optional`s and appending them at the end of a query.

While this should make no difference to `core.logic` it could make a difference in a SPARQL query, so one needs to be aware that those two are not the same. 

### `select`

1. After `conda -> conde` change, we get different results.
2. We want the results to as close to what a SPARQL query would return.

For example, a SPARQL query returns:

```
(def DB1 [(->Triple :john :status :online)
            (->Triple :john :prop1 "A")
            (->Triple :john :prop1 "B")
            (->Triple :john :prop2 :x)
            (->Triple :john :prop2 :y)])

(select [?o ?p ?x]
    [[?o :status ?status]
     (optional [[?o :prop1 ?p]])
     (optional [[?o :prop2 ?x]])]
    DB1)
 
;; result we expect, and what we would get in SPARQL: 
([:john "B" :y] [:john "B" :x] [:john "A" :y] [:john "A" :x])

;; result we get, after `conde` change, but without processing the solutions:
([:john "B" _0] [:john _0 _1] [:john "A" _0] [:john _0 :y] [:john _0 :x])

;; And if we remove the :prop1 relations from DB1, we should get:
([:john _1 :y] [:john _1 :x])

;; but without processing the solutions from `core.logic` we get:
([:john _0 _1] [:john _0 :y] [:john _0 :x])

```

 The changes to `select` macro are cleaning up the unbound variables.

### CODE REVIEW

File: test/grafter/matcha/alpha_test.clj [hunk 1]
```diff
@@ -663,3 +663,32 @@
...
+
+(deftest issue-21-test
+  (testing "Order of `optional`s shouldn't matter."
+    (let [data [[1 :p :a]
+                [1 :p2 :X]
+                [1 :p3 :Z]
+                [3 :q :x]]
+          ;; two 'equal' queries with different ordering of
+          ;; `optional`s
+          result-ab (build [:id ?id]
```
[the remark this answers is not in the document]
Comment (RickMoynihan) [reply test/grafter/matcha/alpha_test.clj:675]: false positive

File: test/grafter/matcha/alpha_test.clj [hunk 2]
```diff
@@ -663,3 +663,32 @@
...
+                           {:id ?id 
+                            :optional-a ?oa
```
[the remark this answers is not in the document]
Comment (RickMoynihan) [reply test/grafter/matcha/alpha_test.clj:677]: false positive

File: test/grafter/matcha/alpha_test.clj [hunk 3]
```diff
@@ -663,3 +663,32 @@
...
+                            :optional-b ?ob}
```
[the remark this answers is not in the document]
Comment (RickMoynihan) [reply test/grafter/matcha/alpha_test.clj:678]: false positive

File: src/grafter/matcha/alpha.clj [hunk 4]
```diff
@@ -198,7 +198,7 @@
      (concat
       (mapv parse-pattern-row requireds)
       (when (seq optionals)
-        [`(l/conda
+        [`(l/conde
```
Comment (RickMoynihan) [inline src/grafter/matcha/alpha.clj:201]: Ok, so I'm assuming that this is the root source of the bug (with the rest of the changes) being to clean up the  implications this has on the results.

To be honest I'd thought we were doing a `l/conde`, and I can't think why this was implemented as a `l/conda`.  Though I looked and it seems we're wrapping the `l/conde` with an `l/conda`, so the `l/conde` I thought we were doing is at a different level.

I've checked in the history and it seems that originally I'd [said `OPTIONAL` was just a `conde`](https://github.com/Swirrl/matcha/issues/11#issuecomment-425566212) and that @andrewmcveigh had [originally implemented it as such here](https://github.com/Swirrl/matcha/pull/16/files#diff-cc4d6676244bf1219ff935b86132a16670d4d9e18fec723907532fcc855c4055R161) but that it was [later wrapped with a `conda`](https://github.com/Swirrl/matcha/commit/ebd019badb615ba3c22041522ae9499b52739252#diff-cc4d6676244bf1219ff935b86132a16670d4d9e18fec723907532fcc855c4055R185) as part of improvements to the parser/validator, and this is what you've changed to a `conde`

Now this comes up I do vaguely recall discussing this in person with @andrewmcveigh and us both feeling it wasn't quite right, but for some reason (probably the reason for this patch) `conde` didn't quite work, and that `conda` at the time seemed to give correct results, though I have very little recollection of the details, and obviously didn't give it enough attention at the time.

I'm also curious if the optionals worked properly prior to the parser/validator changes, and if this was a regression.

File: src/grafter/matcha/alpha.clj [hunk 5]
```diff
@@ -379,25 +439,70 @@
...
+                         (symbol? (first v))
+                         (= #'optional (resolve (first v)))))
+        pruned (walk/postwalk
+                (fn [form]
+                  (if-not (vector? form)
+                    form
+                    (reduce (fn [result item]
+                              (if (optional? item)
+                                (do
+                                  (tap> {:found item})
```
[the remark this answers is not in the document]
Comment (rosado) [reply src/grafter/matcha/alpha.clj]: already removed on a different branch

**Comment by RickMoynihan** [comment]
Thanks a million for this @rosado it's great work! 🙇

**Pull request merged**